code/controllers/subsystem/tts.dm

```diff
@@ -0,0 +1,227 @@
+				turf_source,
+				sound,
+				vol = listening_mob == target? 60 : 85,
+				falloff_exponent = SOUND_FALLOFF_EXPONENT,
+				channel = channel,
+				pressure_affected = TRUE,
+				sound_to_use = sound,
+				max_distance = SOUND_RANGE,
+				falloff_distance = SOUND_DEFAULT_FALLOFF_DISTANCE,
+				distance_multiplier = 1,
+				use_reverb = TRUE
+			)
+
+/datum/controller/subsystem/tts/proc/handle_request(list/entry)
+	var/timeout_time = entry[START_TIME_INDEX] + message_timeout
+	if(timeout_time < world.time)
+		cached_voices -= entry[IDENTIFIER_INDEX]
+		return
+	var/datum/http_request/request = entry[REQUEST_INDEX]
+	request.begin_async()
+	in_process_tts_messages += list(entry)
+
+/datum/controller/subsystem/tts/fire(resumed)
+	if(!tts_enabled)
+		flags |= SS_NO_FIRE
+		return
+
+	if(!resumed)
+		while(length(in_process_tts_messages) < max_concurrent_requests && queued_tts_messages.len > 0)
+			var/list/entry = popleft(queued_tts_messages)
+			handle_request(entry)
+		current_processing_tts_messages = in_process_tts_messages.Copy()
+
+	// For speed
+	var/list/processing_messages = current_processing_tts_messages
+	while(processing_messages.len)
+		var/current_message = processing_messages[processing_messages.len]
+		processing_messages.len--
+		var/datum/http_request/request = current_message[REQUEST_INDEX]
+		if(!request.is_complete())
+			continue
+
+		var/datum/http_response/response = request.into_response()
+		in_process_tts_messages -= list(current_message)
+		average_tts_messages_time = MC_AVERAGE(average_tts_messages_time, world.time - current_message[START_TIME_INDEX])
+		// If it took too long to process, don't bother playing it
+		var/timeout_time = current_message[START_TIME_INDEX] + message_timeout
+		if(response.errored || timeout_time < world.time)
+			cached_voices -= current_message[IDENTIFIER_INDEX]
+			continue
+
+		var/identifier = current_message[IDENTIFIER_INDEX]
+		var/sound/new_sound = new("tmp/tts/[identifier].ogg")
+		play_tts(current_message[TARGET_INDEX], new_sound, current_message[LANGUAGE_INDEX], current_message[LOCAL_INDEX], current_message[MESSAGE_RANGE_INDEX])
+		for(var/extra_target in current_message[EXTRA_TARGETS_INDEX])
+			play_tts(extra_target["target"], new_sound, current_message[LANGUAGE_INDEX], extra_target["local"], extra_target["range"])
+		cached_voices -= identifier
+		if(MC_TICK_CHECK)
+			return
+
+/datum/controller/subsystem/tts/proc/queue_tts_message(target, message, datum/language/language, speaker, filter, local = FALSE, message_range = 7)
+	if(!tts_enabled)
+		return
+
+	var/static/regex/contains_alphanumeric = regex("\[a-zA-Z0-9]", "g")
+	// If there is no alphanumeric char, the output will usually be static, so
+	// don't bother sending
+	if(contains_alphanumeric.Find(message) == 0)
+		return
+
+	var/shell_scrubbed_input = tts_alphanumeric_filter(message)
+	shell_scrubbed_input = copytext(shell_scrubbed_input, 1, 300)
+	var/identifier = sha1(speaker + shell_scrubbed_input + filter)
```

Comment:
does this need to be sha1()? is it to reduce file name sizes for performance reasons?

Reply:
it's to reduce the risk of collisions since md5 is less collision resistant

Reply:
> does this need to be sha1()? is it to reduce file name sizes for performance reasons?

also yeah, since these are recorded to files and having spaces in file names will likely cause issues